pkg/util/scrubber/yaml_scrubber.go

```diff
@@ -0,0 +1,83 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package scrubber
+
+import (
+	"gopkg.in/yaml.v2"
+)
+
+type scrubCallback = func(string, interface{}) (bool, interface{})
+
+func walkSlice(data []interface{}, callback scrubCallback) {
+	for _, k := range data {
+		switch v := k.(type) {
+		case map[interface{}]interface{}:
+			walkHash(v, callback)
+		case []interface{}:
+			walkSlice(v, callback)
+		}
+	}
+}
+
+func walkHash(data map[interface{}]interface{}, callback scrubCallback) {
+	for k, v := range data {
+		if keyString, ok := k.(string); ok {
+			if match, newValue := callback(keyString, v); match {
+				data[keyString] = newValue
+				continue
+			}
+		}
+
+		switch v := data[k].(type) {
+		case map[interface{}]interface{}:
+			walkHash(v, callback)
+		case []interface{}:
+			walkSlice(v, callback)
+		}
+	}
+}
+
+// walk will go through loaded yaml and call callback on every strings allowing
+// the callback to overwrite the string value
+func walk(data *interface{}, callback scrubCallback) {
+	switch v := (*data).(type) {
+	case map[interface{}]interface{}:
+		walkHash(v, callback)
+	case []interface{}:
+		walkSlice(v, callback)
+	}
+}
+
+// ScrubYaml scrubs credentials from the given YAML by loading the data and scrubbing the object instead of the
+// serialized string.
+func (c *Scrubber) ScrubYaml(input []byte) ([]byte, error) {
+	var data *interface{}
+	err := yaml.Unmarshal(input, &data)
+
+	// if we can't load the yaml run the default scrubber on the input
+	if err == nil {
+		walk(data, func(key string, value interface{}) (bool, interface{}) {
+			for _, replacer := range c.singleLineReplacers {
+				if replacer.YAMLKeyRegex == nil {
+					continue
+				}
+				if replacer.YAMLKeyRegex.Match([]byte(key)) {
+					if replacer.ProcessValue != nil {
+						return true, replacer.ProcessValue(value)
+					}
+					return true, defaultReplacement
+				}
+			}
+			return false, ""
+		})
+
+		newInput, err := yaml.Marshal(data)
+		if err == nil {
```

Comment:
If we get an error here, it may be good to show it in the log since it's not an expected situation (afaiu)

Reply:
I agree but the scrubber is on of the building block of the `log` package. So we can't use `log` here.

Reply:
What about fmt.Printf then? Seems like if we end up with an error here we really need to know about it.

Reply:
`fmt.Printf` would not be captured in the logs so we would never saw it.

Reply:
Not in the logs but we would see it in the agent output 🤷